### PR TITLE
Remove filter columns from table scans if they are unused in the remainder of the plan

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR adds an optimization to the `RemoveUnusedColumns` optimizer that removes filter columns from table scans, if the filter columns are unused in the remainder of the plan, for example:
```sql
SELECT i FROM integers WHERE j = 42;
```
We already perform filter pushdown and projection pushdown for this query. However, for this query, the `PhysicalTableScan` operator will currently emit both columns `i` and `j`. A projection will sit on top of the table scan, which will immediately remove `j` from the remainder of the plan.

In general, this is fine. However, when performing remote scans, this means that `j` wil be transferred over the network, causing unnecessary overhead.

This PR removes the columns within the `TableFunction`, so this should no longer be an issue.

### CODE REVIEW

**Comment by hannes** [comment]
CC @nickursa

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good to me.

One comment - the Postgres scanner also supports both projection and filter pushdown, and as such will be affected by this change as well. @hannes correct me if I'm wrong here. 

Could you perhaps add a test that targets this behavior to the Postgres scanner repo, and fix it there as well?

**Comment by lnkuiper** [comment]
Good idea, have no clue how to do anything with out-of-tree extensions yet though. Let's check together in the office tomorrow.

**Comment by lnkuiper** [comment]
I've reworked the API for this so that the projection columns are passed to the `TableFunctionInitInput` instead. This should now also no longer affect the Postgres scanner, as the default value of `filter_prune` is false. The Postgres scanner can make good use of this though if it implements this.

**Review by Mytherin — COMMENTED** [review]
Thanks for the rework! Good idea to add the explicit `filter_prune` setting to opt into this behavior.

Just to ensure this works correctly, could we perhaps add tests that test this specific behavior for the Parquet and Arrow scans?

**Comment by lnkuiper** [comment]
I have added some tests, and I think this is good to go when the CI passes.

**Comment by Mytherin** [comment]
Thanks!

File: tools/pythonpkg/tests/fast/arrow/test_filter_pushdown.py [hunk 1]
```diff
@@ -363,3 +363,16 @@ def test_filter_pushdown_2145(self,duckdb_cursor):
...
+    def test_filter_column_removal(self,duckdb_cursor):
+        if not can_run:
+            return
+        duckdb_conn.execute("CREATE TABLE test AS SELECT range i, range j FROM range(5)")
+        duck_test_table = duckdb_conn.table("test")
+        arrow_test_table = duck_test_table.arrow()
+        duckdb_conn.register("arrow_test_table",arrow_test_table)
+
+        # PR 4817 - remove filter columns that are unused in the remainder of the query plan from the table function
+        query_res = duckdb_conn.execute("EXPLAIN SELECT count(*) from testarrow where a = 100 or b =1").fetchall()
```
Comment (Tishj) [inline tools/pythonpkg/tests/fast/arrow/test_filter_pushdown.py:376]: This test creates `arrow_test_table`, but the query uses `testarrow`
@lnkuiper I'm not sure what the intent here was

**Pull request merged**